### ISSUE TITLE
How do you balance reaction stoichiometry?

### ISSUE THREAD

**Issue by bionicles** [connorcoley/rexgen_direct#6]
Just curious... how could we go through the data file and balance these reactions to make sure there are equal numbers of atoms on both sides?

I want to make simulations of them, but it's hard if they're imbalanced

Many thanks,
bionicles

**Comment by connorcoley** [comment]
The code expects that it will be provided atom-mapped reaction SMILES where all atoms in the product molecule(s) have unique atom map numbers that are present in the reactants. The code will break if there are duplicated atom map numbers (e.g., indicating reactant stoichiometry of two by using those map numbers in the product multiple times) as opposed to duplicated reactants (e.g., indicating reactant stoichiometry by providing two explicit copies of the reactants using different numbering).

This does _not_ mean that the reactions are fully balanced. The datasets we work with rarely contain information about byproducts (e.g., salts, water) and so the code does not expect them. The parent data source can be found [here](https://figshare.com/articles/Chemical_reactions_from_US_patents_1976-Sep2016_/5104873). There is no open-source reaction dataset with truly balanced reactions. [This paper](https://onlinelibrary.wiley.com/doi/full/10.1002/anie.201503890) uses a proprietary set that required hand curation.

**Comment by amrhamedp** [comment]
which code was used for atom mapping to prepare the data set ?

**Comment by connorcoley** [comment]
The dataset was previously mapped using Indigo

**Comment by bionicles** [comment]
If reactions are coded with balanced and matched PDB files then simulating them with NNs is trivial because you could just plug them into the same exact pipeline for proteins, otherwise you need either to balance the reactions (less ideal honestly) or figure out a loss function which handles unbalanced / unmatched IO (super valuable if done abstractly)

we got pretty bogged down in using DNNs for atomistic simulations of chemistry due to permutation invariance issues, and decided to move on to more "boring" stuff (data security) but for future reference to interested folks it could possibly work with a Fused Gromov Wasserstein loss function on the neural network; compare unmatched molecule graphs with transport theory, with adding the "feature distance" between the atoms features as another transport cost to minimize

there is a critical need for alignment-free and permutation invariant loss functions for neural networks which move atoms (ATOM MOVER DISTANCE). some past work uses local kernels for this, which is neat, it would be cool to try multiscale kernels or FGW. If such a loss function existed then you could train 1 neural network to do quantum physics, organic chemistry, and biomolecular engineering. Not sure if this is worth it in the era of systems and synthetic biology (why make a molecule when you can make a circuit?) but would be cool to look at. I'd like to circle back to this in a few months and use PyTorch + Python Optimal Transport if anyone's keen, I already made the OpenAI gym env for PyMol, but Tensorflow 2.0 wasnt as super fun as I hoped

https://tvayer.github.io/materials/Titouan_Marseille_2019.pdf

thanks for your work, feel free to keep this open or close it
